### ISSUE TITLE
5.4 breaks multiple hostnames for website

### ISSUE THREAD

**Issue by isecchin** [hyn/multi-tenant#750]
__Description__

First of, thanks for the amazing package ! I've been using for awhile and it saved me a lot of time !

Migrating to Laravel 5.8 and consequently to multi-tenant 5.4, I've been experiencing an unexpected behavior with my hostnames, though.

Basically I have 2 hostnames (one for internal use and one for external, using a NOIP domain) for the same Website.

The internal hostname works just fine with a fake domain, registered in the `etc/hosts` file, but since the upgrade, I cannot access it from the external domain.

Example:

1. sub.domain.com
2. sub.noip.com

Access through the sub.domain.com on my local machine works fine, but trying to access it from the sub.noip.com, I get redirected to the sub.domain.com that does not exist, and can't access the system (which I could on 5.3).

---

__Actual behavior__

Can't access the Website by secondary Hostnames.

__Expected behavior__

Be able to access the Website by all its Hostnames.

---
__Information__

- hyn/multi-tenant version: 5.4
- laravel version: 5.8

---

__Debugging__

I've been playing around with the source code and trying to check all the changes made from one version to the other and I think I've tracked down the problem to the `Hyn\Tenancy\Listeners\URL\UpdateAppUrl` class.

More specifically, on line 40:

```php
$hostname = $event->hostname ?? $event->website->hostnames->first();
```

So basically it's just grabbing the first hostname registered for the website, which I believe shouldn't be true, I guess we should first verify if the request's host matches any of the hostname's FQDN, so basically just do something like this:

```php
$hostname = $event->hostname
    ?? $event->website->hostnames->firstWhere('fqdn', optional(request())->getHost())
    ?? $event->website->hostnames->first();
```

I'm not sure if the current behavior is actually the intended one or if the above solution would be ok (it seems to work alright), if so I could also open a PR.

**Comment by isecchin** [comment]
Actually, after further debugging, I checked that if a different port is used (other than the default 80), the issue is still present, and I think it is because of the `URL::forceRootUrl()` function.

So I think we probably need to update the URL as well with the port number, if it is present, so something like this, perhaps ?

```php
if ($hostname) {
    $url = sprintf('%s://%s', $scheme, $hostname->fqdn);

    config([
        'app.url' => $url
    ]);

    URL::forceRootUrl(optional(request())->getPort()
        ? ($url . ':' . request()->getPort())
        : $url
    );
}
```

**Comment by ArlonAntonius** [comment]
Are you using ports in FQDN?
Why are you using a different port than port 80 for something?

**Comment by isecchin** [comment]
Hi @ArlonAntonius, sorry, maybe I wasn't clear with my last comment.

No, I don't register the port with my FQDN, the registries in the DB are stored only with the FQDN (no ports), but before version 5.4 I could use a different port to access a hostname without any issues.

Ideally I wouldn't want to use a different port as this is just inconvenient most of the times, but unfortunately some internet providers block the 80 port, so I have to sometimes rely on alternative ports such as 8080 to access the domain externally.

I know this is just an issue for probably a development environment, but I think some other services may also require to run on a different port (I'm not sure, I might be wrong about this), so I think if we always force a 'redirect' to port 80 (or 443) when a request might come from a different port, we might break something else.

**Comment by bkintanar** [comment]
hmm.. I use port 8080 and port 8443 on a daily basis without problems. I need to check on this more.

**Comment by isecchin** [comment]
Well, I'm not sure if my case is too specific (probably so hahaha), but the problem presented itself basically for having 2 (or more) hostnames linked to the same website, with one being used on port 80 (locally with the fake domain) and the other being used on port 8080 (with a NO-IP domain).

Although the port caused an issue with the code I shared on my first comment (which I could patch with the code from my second comment), I think the main problem is present nonetheless, because it is always redirecting to the first hostname registered for the website instead of searching for the right one.

**Comment by ArlonAntonius** [comment]
@isecchin What do you mean with "Instead of searching for the right one"?
We're simply booting with the first hostname found for the website, which should be the right one?

**Comment by isecchin** [comment]
@ArlonAntonius I meant based on the request's URL (as per my code example).

So, for instance, I have two hostnames registered for the same website (`sub.domain.com` and `sub.noip.com`).

The way it is implemented right now, whenever I try to access the second domain (`sub.noip.com`) I get redirected to the first one (`sub.domain.com`).

**Comment by ArlonAntonius** [comment]
I'm not sure if this is actual behavior that we should change, I see space for it, but just not sure on it as this is quite a specific use case.
Most people use only 1 Hostname for a tenant anyways.

**Comment by isecchin** [comment]
I partly agree with you on this.

Although I do think most people use Websites as a 1:1 relationship with Hostnames, I think that there are some use cases for having more hostnames for the same website.

If they were to be a 1:1 relationship only, would it be really necessary to be two different things ? Why not just a Website with a FQDN (and any other attributes that are currently on the hostnames) ?

Anyways, I don't think verifying the actual hostname would imply in any performance drop or something like it and it's really a non-invasive solution. I do think though, that if we don't go in this direction to "fix" this, we should at least document this as a breaking change from 5.3 to 5.4, saying that a 1:N relationship is no longer supported, so people would be aware of this before upgrading. For me it wasn't such a big deal since I'm still on the development stage, but if there are others out there that for some reason use more than one hostname for a website and are already with their product in production, this would cause an unexpected behavior.

**Comment by ametad** [comment]
I use two domain names per website: 

*  foo.test
*  www.foo.test

Above is also the order in which the hostnames are created.  If I now go to www.foo.test, all links are generated with `foo.test` as root. As *foo.test* is of course the first().

I want to use the `www.foo.test` domain, so I redirect *foo.test* to *www.foo.test*:

![image](https://user-images.githubusercontent.com/1582541/54594685-7b0fd980-4a31-11e9-9274-37a4d65bff76.png)

Outcome: all links are generated to the wrong domain, and upon entering the user is redirected back to essentially the domain where it came from.

**Comment by ametad** [comment]
PS. Note the `redirect_to` attribute of a Hostname. This is build in functionality. 

`\Hyn\Tenancy\Middleware\HostnameActions` is redirecting the user accordingly.

**Comment by aaronfahey** [comment]
I believe I am experiencing the same issue here.

We have many hostnames per website, which are generally used to distinguish countries. @isecchin I definitely think 1:N hostnames should be supported. Your code seems like it will do the trick. @luceos are we able to open a PR with the code highlighted in the issue description?

**Comment by ArlonAntonius** [comment]
@MrFahey Are you referring to the first code in the initial issue or to the port-related code?

**Comment by aaronfahey** [comment]
@ArlonAntonius sorry should have been more specific - I am referring to this:

`$hostname = $event->hostname
    ?? $event->website->hostnames->firstWhere('fqdn', optional(request())->getHost())
    ?? $event->website->hostnames->first();`

in `Hyn\Tenancy\Listeners\URL\UpdateAppUrl`

**Comment by ametad** [comment]
Perhaps a fallback for when no hostname is known? This would introduce a new feature..

**Comment by luceos** [comment]
Feels like two different things are being discussed here. One relates to using ports in requested hostnames and forcing `app.url`, the other relates to using redirect_to.

@MrFahey Aaron are you sure your issue relates to this, because 1:N is natively supported and your example hostnames don't use a port at all..

**Comment by isecchin** [comment]
Hi @luceos, I can't say for sure about @MrFahey issue, but I believe it's also related to mine.
The problem is that 1:N is not supported anymore since the 5.4 release, which updated the `Hyn\Tenancy\Listeners\URL\UpdateAppUrl`.

The port issue is also related to the same update of forcing the `app.url` and the `forceRootUrl()` call. Before version 5.4, the 1:N relationship was indeed supported and worked for both cases described in this issue (many hostnames and different ports). During the update though, it is forcing now to always get the first hostname, no matter what hostname was actually typed by the user, and the `forceRootUrl()` call just ignores the port.

**Comment by ametad** [comment]
> Feels like two different things are being discussed here. One relates to using ports in requested hostnames and forcing `app.url`, the other relates to using redirect_to.

Yes but if I may say so... It is related to the issue that 1:N is not working anymore. 

The point I want to make with my example before, is that I also think it is a good idea to support multiple hostnames (1:N).

**Comment by londoh** [comment]
Hi, 

I've upgraded to 5.4.4 and l5.8 and am also hitting this issue.
I also have multiple hostnames listed per website. Issue appears where app.url is getting first() hostname which is incorrect.

As to the debate as to whether multiple hostnames should be supported - There are comments in config as follows:
```
         * Automatically update the app.url configured inside Laravel to match
         * the tenant FQDN whenever a hostname/tenant was identified.
```
So its expected that app.url will be set to fqdn whenever. 
And indeed it did work as expected (at least in 5.3.1) but in my case after this upgrade app.url no longer matches actual tenant fqdn unless by chance fqdn == $hostnames->first()
So this looks like a breaking change and a bug

I agree with others that its being generated in UpdateAppUrl.
In my app UpdateAppUrl::force is called 6 times. Logging event class gives this:
```

[2019-06-08 17:50:16] local.INFO: Event: Hyn\Tenancy\Events\Websites\Identified  
[2019-06-08 17:50:16] local.INFO: Event: Hyn\Tenancy\Events\Websites\Switched  
[2019-06-08 17:50:17] local.INFO: Event: Hyn\Tenancy\Events\Websites\Identified  
[2019-06-08 17:50:17] local.INFO: Event: Hyn\Tenancy\Events\Websites\Switched
[2019-06-08 17:50:17] local.INFO: Event: Hyn\Tenancy\Events\Websites\Identified  
[2019-06-08 17:50:17] local.INFO: Event: Hyn\Tenancy\Events\Websites\Switched 
```
These $event 's dont arrive at UpdateAppUrl with same data. But...
after 'Websites\Identified' event app.url is correct
after 'Websites\Switched' event app.url is incorrectly given as first()

maybe this helps, I will try and look into it further and report back.

regards

l.


- hyn/multi-tenant version: 5.4.4
- laravel version: 5.8.
- database driver and version: mariadb 10.3.15
- webserver software and version: apache 2.4.6
- php version: 7.2.19
config as per: https://github.com/tenancy/multi-tenant/issues/613

**Comment by fletch3555** [comment]
This is likely because [this query](https://github.com/tenancy/multi-tenant/blob/5.x/src/Listeners/URL/UpdateAppUrl.php#L40) doesn't specify order, so when there are multiple results, the object returned is arbitrary.  There really isn't a good solution for this, except perhaps to somehow identify the "preferred" hostname

**Comment by londoh** [comment]
this seems to work for me at this time:
```
    public function force($event)
    {
        if (config('tenancy.hostname.update-app-url', false)) {
            $scheme = optional(request())->getScheme() ?? parse_url(config('app.url'), PHP_URL_SCHEME);

            /** @var Hostname $hostname */
            //$hostname = $event->hostname ?? $event->website->hostnames->first();
            $hostname = $event->hostname ?? $event->website->hostnames
                                                    ->where('fqdn', request()->getHost())
                                                    ->first();
            if ($hostname) {
                $url = sprintf('%s://%s', $scheme, $hostname->fqdn);

                config([
                    'app.url' => $url
                ]);

                URL::forceRootUrl($url);
            }
        }
    }
```
... but I've been away from multi-tenant for a while so if I doing something dumb here please let me know!

regards

l.

**Comment by londoh** [comment]
hmmm that doesnt work in cli.
so here's another temp kludge:
```
            /** @var Hostname $hostname */
            $hostname = $event->hostname ??
                        $event->website->hostnames
                        ->where('fqdn', request()->getHost())->first() ??                
                        $event->website->hostnames->first();
```

**Comment by drbyte** [comment]
I will assert that the workaround posted by @isecchin  in first post above (https://github.com/tenancy/multi-tenant/issues/750#issue-421311480) correctly resolves the stated problem. I will re-state it again, for clarity: when a website has multiple hostnames, this change will correctly "keep" the URL that the visitor came in on, instead of switching unexpectedly to the "first found by Eloquent".
For me this is critical.
@luceos will you accept a PR with that fix? 

IMO any concerns about "other ports" should be handled separately.
Similarly, there are already other discussions about multiple firing of events. I wouldn't muddy this issue with further discussion about those.

**Comment by londoh** [comment]
I agree with @drbyte. And thanks for pointing out the the solution already posted by @isecchin - I'd overlooked it. Apologies to all for prolonging the thread.

**Comment by luceos** [comment]
@drbyte Yes Chris please do PR I'll check it out asap, thanks.

**Comment by isecchin** [comment]
@drbyte I agree with you that the port problem is another issue and could be handled separately, the only reason I brought it up in this thread was because on my specific case both problems appeared together since the two hostnames I had responded in different ports as well and the commit that broke the hostnames also broke the ports.

What I did before here (on [this comment](https://github.com/tenancy/multi-tenant/issues/750#issuecomment-473258854)) is a bit of a hack and it does not work as expected.

I've had some other implementations locally to make it work for my use case, but all of them seemed like too much work for something so simple, and I ended up with this instead:

**Hyn\Tenancy\Listeners\URL\UpdateAppUrl**

```php
<?php

/*
 * This file is part of the hyn/multi-tenant package.
 *
 * (c) Daniël Klabbers <daniel@klabbers.email>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @see https://laravel-tenancy.com
 * @see https://github.com/hyn/multi-tenant
 */

namespace Hyn\Tenancy\Listeners\URL;

use Hyn\Tenancy\Contracts\CurrentHostname;
use Hyn\Tenancy\Contracts\Hostname;
use Hyn\Tenancy\Events\Websites\Identified;
use Hyn\Tenancy\Events\Websites\Switched;
use Illuminate\Contracts\Events\Dispatcher;
use Illuminate\Support\Facades\URL;

class UpdateAppUrl
{
    public function subscribe(Dispatcher $events)
    {
        $events->listen([Identified::class, Switched::class], [$this, 'force']);
    }

    /**
     * @param Identified|Switched $event
     */
    public function force($event)
    {
        if (config('tenancy.hostname.update-app-url', false)) {
            $scheme = optional(request())->getScheme() ?? parse_url(config('app.url'), PHP_URL_SCHEME);

            /** @var Hostname $hostname */
            $hostname = $event->hostname
                ?? $event->website->hostnames->firstWhere('fqdn', optional(request())->getHost())
                ?? $event->website->hostnames->first();

            if ($hostname) {
                $url = sprintf('%s://%s', $scheme, $hostname->fqdn);

                config([
                    'app.url' => $url
                ]);
            }
        }
    }
}
```

So basically I did what I suggested on the issue description itself (to fix the 1:N hostname problem) and removed the `URL::forceRootUrl()` call (this fixed the problem with the ports).

I don't know exactly what was the motivation to include this forceRootUrl call (since it didn't exist on previous versions), maybe it was to fix another issue, I don't know, if someone could clarify on that it would be nice.

If you think there's no implications on removing it, I could open a PR with what I have now.

**Comment by drbyte** [comment]
@luceos or @ArlonAntonius can you confirm whether removing `URL::forceRootUrl()` would be problematic? Or under which conditions it would be needed?
https://github.com/tenancy/multi-tenant/blob/81c52238225d95dc5ea32913efe9ba507d048fac/src/Listeners/URL/UpdateAppUrl.php#L49

**Comment by ametad** [comment]
@isecchin I think I can provide a little inside of why the `forceRootUrl` is implemented some time ago. I had some problems with generating 'urls' in a queued job and filed an issue here at Gitlab.

The root url needs to be re-set at runtime on command line interface (cli) to generate an url (e.g. `url('/home')` in an Email), when the worker handles a queued job that has Tenant context. The thing is that on the command line there is no request to refer to, so no current url is available. This introduces the need to find related FQDN and set it as the root url. 

There is another scenario where you want to have a new url root set, that is when you switch to another tenant. I need this because of a mailing that generates links to many different tenant domains. Like this:

```php
$website = \Hyn\Tenancy\Repositories\WebsiteRepository::findById($id)

// multi-tenant 5.3
Mail::to($email)->send((new Invoice($this))->onTenant($website));

// multi-tenant 5.4
\Hyn\Tenancy\Facades\TenancyFacade::tenant($website);
Mail::to($email)->send(new Invoice($this));

```

**Comment by drbyte** [comment]
@ametad That makes sense.

@isecchin let's still keep the two matters separate (1.hostname, 2.port), even though they're in the same file.
Since you started this discussion I'm content to let you make the PR if you have time today. If not, I'll do it this evening.

**Comment by ametad** [comment]
@drbyte Thank you.

**Comment by isecchin** [comment]
@ametad thanks for the clarification !

@drbyte yes, I agree, if there was a reason to use the forceRootUrl and it was to fix another issue, we should proceed with more caution on that and separate the matters, I'll create a PR with the hostnames' fix only and open a new, separate, issue about the ports so we can discuss and debug further for another possible solution that doesn't break other features.
